WordPress/gutenberg
# [Mobile] Remove Keyboard Aware Flatlist

Fixes https://github.com/WordPress/gutenberg/issues/40029
Fixes https://github.com/wordpress-mobile/gutenberg-mobile/issues/4894

**Related PRs:**
- https://github.com/wordpress-mobile/gutenberg-mobile/pull/5533
- https://github.com/wordpress-mobile/WordPress-iOS/pull/20262
- https://github.com/wordpress-mobile/WordPress-Android/pull/18082

## What?
This PR fixes a high-priority bug on iOS where the scroll would jump to the top and then scroll back to the currently focused text input.

## Why?
Because it affects the basic writing flow, and after several attempts to try to workaround that library I went to a different route by implementing it within the project and stop using our fork for [react-native-keyboard-aware-scroll-view](https://github.com/wordpress-mobile/react-native-keyboard-aware-scroll-view/tree/wp-fork).

## How?

### BlockList

- It removes the `OnCaretVerticalPositionChange` context and provider since it won't be used anymore.
- Simplifies the number of properties set for the `KeyboardAwareFlatList` component.
- Fixes the logic for `isFloatingToolbarVisible` since it was not returning the correct value.

### RichText (Block-editor)

- Removes `useNativeProps` since it won't be used anymore.

### KeyboardAwareFlatList

This component got a full refactor, it will leverage on the `InputState` from `RCTAztecView` for different cases by using new hooks:

* **useKeyboardOffset**: Hook that adds Keyboard listeners to get the offset space when the keyboard is opened, taking into account focused AztecViews.
* **useScrollToTextInput**:  Hook that provides the `scrollToTextInputOffset` function to scroll to the currently focused TextInput depending on where the caret is placed.
* **useTextInputCaretPosition**: Hook that listens to caret changes from AztecView TextInputs.
* **useTextInputOffset**: Hook that calculates the currently focused TextInput's current caret Y coordinate position within the ScrollView.

### VisualEditor

- Removes the `isAutoScrollEnabled` state value which is no longer needed.

### AztecInputState

- It introduces new logic to store the current caret's Y coordinate, as well as a new listener to trigger a notification when this value changes.

### AztecView

- Removes usage of `onCaretVerticalPositionChange` and now it will only set the value in `AztecInputState`.

### E2E Tests

- Improves the `getTitleElement` helper and the block insertion canary test.

## Testing Instructions

## Test case 1

- Open the app using the builds linked to this PR.
- Create a new post
- Add a Paragraph block
- Tap on the return key of the keyboard
- Repeat the same step until you reach the end of the editor
- Keep adding paragraph blocks and **expect** the scroll **to not** scroll up to the title

## Test case 2

- Open the app using the builds linked to this PR.
- Open a post that has at least 12 paragraph blocks
- Tap on the very last one and expect the scroll to move and focus on the caret's position
- Add more text and expect the scroll to keep the focus on the caret's position if it's right at the bottom of the editor.

### Testing Instructions for Keyboard
N/A

## Screenshots or screencast 

## Test case 1

Before | After
-|-
<video src="https://user-images.githubusercontent.com/4885740/224366271-742221db-6c11-4d58-8c01-d777e3e32dc7.mov"/> | <video src="https://user-images.githubusercontent.com/4885740/224365220-78335bb8-cde9-4586-8634-a475a8a13424.mov"/>

## Test case 2

Before | After
-|-
<video src="https://user-images.githubusercontent.com/4885740/224366532-59457c9c-bfe0-4b5e-b673-c7871feffa13.mov"/> | <video src="https://user-images.githubusercontent.com/4885740/224365392-113aa889-ce5b-4ef4-b416-fa89f3058ab9.mov"/>
